pint/pandas_interface/pint_array.py

```diff
@@ -0,0 +1,609 @@
+    # numpy array
+    type = build_quantity_class(_DEFAULT_REGISTRY, force_ndarray=True)
+    # # AS: I'm not sure that does force it as an ndarray.
+    # # Trying the below as running into registry issues
+    # type = _Quantity
+    name = 'pint'
+
+    @classmethod
+    def construct_array_type(cls, type_str='pint'):
+        if type_str is not cls.name:
+            raise NotImplementedError
+        return PintArray
+
+    @classmethod
+    def construct_from_string(cls, string):
+        if string == cls.name:
+            return cls()
+        else:
+            raise TypeError("Cannot construct a '{}' from "
+                            "'{}'".format(cls, string))
+
+
+class PintArray(ExtensionArray, ExtensionOpsMixin):
+    _dtype = PintType
+
+    def __init__(self, values, dtype=None, copy=False):
+        if isinstance(values, _Quantity):
+            self._dtype.type = type(values)
+            assert self._dtype.type._REGISTRY == values._REGISTRY
+        self._data = self._coerce_to_pint_array(values, dtype=dtype, copy=copy)
+
+    def _coerce_to_pint_array(self, values, dtype=None, copy=False):
+        if isinstance(values, self._dtype.type):
+            return values
+
+        if is_list_like(values):
+            if all(is_bool(v) for v in values):
+                # known bug in pint https://github.com/hgrecco/pint/issues/673
+                raise TypeError("Invalid magnitude for {}: {}"
+                                "".format(self._dtype.type, values))
+
+            for i, v in enumerate(values):
+                if isinstance(v, self._dtype.type):
+                    continue
+                else:
+                    values[i] = v * self._find_first_unit(values)
+
+            units = set(v.units for v in values)
+            if len(units) > 1:
+                # need to work out a way to test this
+                raise TypeError("The units of all quantities are not the same"
+                                " for input {}".format(values))
+
+            magnitudes = [v.magnitude for v in values]
+
+            return self._dtype.type(magnitudes, values[0].units)
+
+        return NotImplementedError
+
+    def _find_first_unit(self, values):
+        for v in values:
+            if isinstance(v, self._dtype.type):
+                return v.units
+
+        return self._dtype.type(1).units
+
+    def __getitem__(self, item):
+        # type (Any) -> Any
+        """Select a subset of self.
+        Parameters
+        ----------
+        item : int, slice, or ndarray
+            * int: The position in 'self' to get.
+            * slice: A slice object, where 'start', 'stop', and 'step' are
+              integers or None
+            * ndarray: A 1-d boolean NumPy ndarray the same length as 'self'
+        Returns
+        -------
+        item : scalar or PintArray
+        """
+        if is_integer(item):
+            return self._data[item]
+
+        return type(self)(self._data[item])
```

Comment:
Why type(self) and not self.__class__ ?

Reply:
No idea, we have lots of oddities in our style (which could be fixed with a linter, see discussion point above?). Changed now.